core/src/sample_performance_service.rs

```diff
@@ -79,7 +79,7 @@ impl SamplePerformanceService {
                 }
             }
 
-            sleep(Duration::from_millis(SLEEP_INTERVAL));
+            sleep(SLEEP_INTERVAL);
         }
```

Comment:
The `now` variable name is wrong.  It is not "now" - it is the time when the last aggregation happened.  How about refactoring the whole loop like this:

```suggestion
    pub fn run(
        bank_forks: Arc<RwLock<BankForks>>,
        blockstore: Arc<Blockstore>,
        exit: Arc<AtomicBool>,
    ) {
        let mut snapshot = StatsSnapshot::from_forks(&bank_forks);

        let mut last_sample_time = Instant::now();
        while !exit.load(Ordering::Relaxed) {
            let now = Instant::now();
            if now.duration_since(last_sample_time) < SAMPLE_INTERVAL {
                sleep(Duration::from_millis(SLEEP_INTERVAL));
                continue;
            }

            last_sample_time = now;
            update_performance_stats(&mut snapshot, &bank_forks, &blockstore);
        }
    }
```

And doing the update in a separate helper:
```rust
fn update_performance_stats(
    snapshot: &mut StatsSnapshot,
    bank_forks: &RwLock<BankForks>,
    blockstore: &Blockstore,
) {
    let new_snapshot = StatsSnapshot::from_forks(&bank_forks);

    let (num_transactions, num_non_vote_transactions, num_slots) =
        new_snapshot.diff_since(&snapshot);

    // Store the new snapshot to compare against in the next iteration of the loop.
    *snapshot = new_snapshot;

    let perf_sample = PerfSampleV2 {
        // Note: since num_slots is computed from the highest slot and not the bank
        // slot, this value should not be used in conjunction with num_transactions or
        // num_non_vote_transactions to draw any conclusions about number of
        // transactions per slot.
        num_slots,
        num_transactions,
        num_non_vote_transactions,
        sample_period_secs: elapsed.as_secs() as u16,
    };

    let highest_slot = snapshot.highest_slot;
    if let Err(e) = blockstore.write_perf_sample(highest_slot, &perf_sample) {
        error!("write_perf_sample failed: slot {:?} {:?}", highest_slot, e);
    }
}
```

Reply:
Some good points; I made a few tweaks:
- Rename `now` to `last_sample_time`
- Replace `loop` with `while` and check exit flag in the while loop predicate
- run() doesn't need to be `pub`

I like the thinking with adding another helper, but opted to skip it here for several reasons:
- The loop is relevant to how we capture samples. Namely, the state from previous iteration we use to calculate a "delta" as well as the actual elapsed time to set `sample_period_secs`
- I feel like the function is already pretty manageable as is; 33 lines including comments/braces/whitespace

Reply:
> Some good points; I made a few tweaks:
> 
>     * Rename `now` to `last_sample_time`
> 
>     * Replace `loop` with `while` and check exit flag in the while loop predicate
> 
>     * run() doesn't need to be `pub`

:+1: 

> I like the thinking with adding another helper, but opted to skip it here for several reasons:
> 
>     * The loop is relevant to how we capture samples. Namely, the state from previous iteration we use to calculate a "delta" as well as the actual elapsed time to set `sample_period_secs`

`&mut` explicitly shows we are changing the value.

>     * I feel like the function is already pretty manageable as is; 33 lines including comments/braces/whitespace

I do not like when things are deeply nested.  `if` is covering almost the whole body of the `while` loop.  So the indentation it adds, in my mind, is "noise".

---

In any case, it is an improvement over what was here before.
Feel free to merge it in, or change it a bit more :P